### ISSUE TITLE
Spotube

### ISSUE THREAD

**Issue by quantenzitrone** [NixOS/nixpkgs#200120]
**Project description**
Spotube is a Flutter based lightweight spotify client. It utilizes the power of Spotify & Youtube's public API & creates a hazardless, performant & resource friendly User Experience.

**Metadata**

* homepage URL: https://spotube.netlify.app/
* source URL: https://github.com/KRTirtho/spotube
* license: BSD-4-Clause License
* platforms: linux, maybe darwin (there is a macOS release)

**Comment by partiallyordered** [comment]
A work in progress:
```nix
  spotube = pkgs.stdenv.mkDerivation rec {
    version = "2.7.1";
    pname = "spotube";
    description = ''
      Open source Spotify client that doesn't require Premium nor uses Electron! Available for both
      desktop & mobile!
    '';
    buildInputs = with pkgs; [
      glib at-spi2-atk harfbuzz cairo gtk3-x11 gst_all_1.gstreamer gst_all_1.gst-plugins-base
    ];
    nativeBuildInputs = [ pkgs.autoPatchelfHook ];
    src = pkgs.fetchzip {
      url = "https://github.com/KRTirtho/spotube/releases/download/v${version}/spotube-linux-${version}-x86_64.tar.xz";
      sha256 = "0kb2vjfyw8rr1xlha66fw4m3b7vl9lrbkafjg7z6dhpjkyrx7wmm";
      stripRoot = false;
    };
    installPhase = ''
      mkdir -p $out
      cp -R $src $out/bin
    '';
    postFixup = ''
      find $out/bin/lib -type f | \
        xargs -I{} basename {} | \
        xargs -I{} patchelf $out/bin/spotube --add-needed {}
      patchelf $out/bin/spotube --add-rpath '$ORIGIN/lib'
    '';
  };
```
Produces the following output when running the `spotube` binary:
```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: MissingPlatformDirectoryException(Unable to get application documents directory)
#0      getApplicationDocumentsDirectory (package:path_provider/path_provider.dart:122)
<asynchronous suspension>
#1      HiveX.initFlutter (package:hive_flutter/src/hive_extensions.dart:12)
<asynchronous suspension>
#2      QueryClient.initialize (package:fl_query/src/core/client.dart:345)
<asynchronous suspension>
#3      main (package:spotube/main.dart:73)
<asynchronous suspension>
```
This might actually be a problem specific to my system configuration. I _think_ it's looking for `XDG_DOCUMENTS_DIR` in environment but I ran out of time to investigate: https://github.com/flutter/packages/blob/746750ecef7852548b038a5ed1a319b7d1d65173/packages/path_provider/path_provider_linux/lib/src/path_provider_linux.dart#L71

**Comment by Quantenzitrone** [comment]
for me it crashes when trying to play something:
```
$> result/bin/spotube
AudioPlayers: Not all elements could be created.

(spotube:789888): GStreamer-CRITICAL **: 08:05:44.866: gst_element_query_duration: assertion 'GST_IS_ELEMENT (element)' failed
AudioPlayers: Could not query current duration.
flutter: GSettings schema org.gnome.desktop.wm.preferences not installed

** (spotube:789888): CRITICAL **: 08:05:46.922: void synchronize_lock_states_loop_body(gpointer, gpointer, gpointer): assertion 'self->caps_lock_state_logic_inferrence != kStateLogicUndecided' failed

** (spotube:789888): CRITICAL **: 08:05:49.687: void synchronize_lock_states_loop_body(gpointer, gpointer, gpointer): assertion 'self->caps_lock_state_logic_inferrence != kStateLogicUndecided' failed
flutter: GSettings schema org.gnome.desktop.wm.preferences not installed
flutter: GSettings schema org.gnome.desktop.wm.preferences not installed
flutter: GSettings schema org.gnome.desktop.wm.preferences not installed
flutter: GSettings schema org.gnome.desktop.wm.preferences not installed
flutter: GSettings schema org.gnome.desktop.wm.preferences not installed
flutter: GSettings schema org.gnome.desktop.wm.preferences not installed

(spotube:789888): GStreamer-CRITICAL **: 08:06:11.585: gst_element_set_state: assertion 'GST_IS_ELEMENT (element)' failed
AudioPlayers: Unable to set the pipeline to the paused state.

(spotube:789888): GStreamer-CRITICAL **: 08:06:11.585: gst_element_query_duration: assertion 'GST_IS_ELEMENT (element)' failed
AudioPlayers: Could not query current duration.

(spotube:789888): GStreamer-CRITICAL **: 08:06:11.615: gst_element_query_duration: assertion 'GST_IS_ELEMENT (element)' failed
AudioPlayers: Could not query current duration.

(spotube:789888): GStreamer-CRITICAL **: 08:06:14.550: gst_element_query_duration: assertion 'GST_IS_ELEMENT (element)' failed
AudioPlayers: Could not query current duration.

(spotube:789888): GStreamer-CRITICAL **: 08:06:14.580: gst_element_query_duration: assertion 'GST_IS_ELEMENT (element)' failed
AudioPlayers: Could not query current duration.

(spotube:789888): GStreamer-CRITICAL **: 08:06:14.580: gst_element_set_state: assertion 'GST_IS_ELEMENT (element)' failed

(spotube:789888): GLib-GObject-CRITICAL **: 08:06:14.581: g_object_set: assertion 'G_IS_OBJECT (object)' failed
fish: Job 1, 'result/bin/spotube' terminated by signal SIGSEGV (Address boundary error)
```

**Comment by Ultra980** [comment]
any progress on this?

**Comment by Steinhagen** [comment]
I also tried to compile this from source, but the build fails:
https://github.com/Steinhagen/nixpkgs/commit/f33e6aa7b8bc14e597afc3c8c654f89493b791fc
To conclude, no useful progress to have this running.

**Comment by TomaSajt** [comment]
I also attempted to package this with the recently merged modification to `buildFlutterApplication`, which now reads the pubspec from a converted pubspec.lock.json file

This is how far I've got:
https://github.com/TomaSajt/nixpkgs/blob/spotube/pkgs/by-name/sp/spotube/package.nix

I got one of the upstream issues solved by manually downloading [a tarball dependency](https://github.com/media-kit/media-kit/blob/main/libs/linux/media_kit_libs_linux/linux/CMakeLists.txt) to a specific directory.

Sadly, I couldn't patch the next one, because that wastrying to  [download a file](https://github.com/KRTirtho/metadata_god/blob/main/packages/metadata_god/linux/CMakeLists.txt) into the efemeral directory of flutter, and it said that it wasn't a writable location (not that it could download anything)
I tried to pre-create the directory, but it didn't seem to change anything

this dependency is published by the same GitHub user as spotube, maybe we could open an issue there related to this...

**Comment by TomaSajt** [comment]
Okay, so I got the tarball dependencies working by using the new `customSourceBuilders` mechanism, which is pretty cool.

Sadly, it turns out the program only supports flutter `3.10.*`, which is lower than the lowest flutter version currently on nixpkgs.
Also, other than the flutter version, the `.env` system would need to be patched too to only require it at runtime.

Here's the current code: https://github.com/TomaSajt/nixpkgs/blob/spotube/pkgs/by-name/sp/spotube/package.nix

**Comment by Steinhagen** [comment]
Probably a ticket should be raised in upstream regarding the flutter support.

**Comment by TomaSajt** [comment]
> Probably a ticket should be raised in upstream regarding the flutter support.

Perhaps, but AFAIK the are aware of it. I got the info from a reply of the repo owner in one of their GitHub Discussions.

**Comment by nephalemsec** [comment]
for nix-darwin users:
```nix
{ stdenv, fetchurl, lib, makeWrapper, undmg }:

  stdenv.mkDerivation rec {
    pname = "spotube";
    version = "3.4.0";
    appname = "Spotube";
    filename = "Spotube-macos-universal";
    meta = with lib; {
      description =
        "An open source, cross-platform Spotify client compatible across multiple platforms
          utilizing Spotify's data API and YouTube (or Piped.video or JioSaavn) as an audio source,
          eliminating the need for Spotify Premium";
      homepage = "https://github.com/KRTirtho/spotube";
      downloadPage = "https://github.com/KRTirtho/spotube/releases";
      license = licenses.bsd3;
    };

    src = fetchurl {
      url =
        "https://github.com/KRTirtho/spotube/releases/download/v${version}/${filename}.dmg";
      hash = "sha256-dif2ec37QL9FVPGgqqvtB2zRX6P3NCth5FFnpASR/As=";
    };

    icon = fetchurl {
      url = "https://raw.githubusercontent.com/KRTirtho/spotube/nightly/assets/spotube-logo.png";
      hash = "sha256-gyzP3ynzuZFN0osmEAjwGfYmbfVxE9Nm2fLJ8q8G6BE=";
    };
  
    sourceRoot = "${appname}.app";
    nativeBuildInputs = [ makeWrapper undmg ];
    installPhase = ''
      runHook preInstall
      mkdir -p $out/{Applications/${appname}.app,bin}
      cp -R . $out/Applications/${appname}.app
      makeWrapper $out/Applications/${appname}.app/Contents/MacOS/${appname} $out/bin/${pname}
      runHook postInstall
    '';
  }
```
Works perfectly on m1 mac.

**Comment by TomaSajt** [comment]
For some reason I didn't try patching the binary distribution originally, so I gave it a try...
And it seems to work!
This is for `linux-x86_64`
```nix
{ lib
, stdenv
, fetchurl

, autoPatchelfHook
, dpkg
, makeWrapper
, wrapGAppsHook

, libappindicator
, libnotify
, libsecret
, mpv-unwrapped
, xdg-user-dirs
}:

stdenv.mkDerivation (finalAttrs: {
  pname = "spotube";
  version = "3.4.0";

  src = fetchurl {
    url = "https://github.com/KRTirtho/spotube/releases/download/v${finalAttrs.version}/Spotube-linux-x86_64.deb";
    hash = "sha256-XBciTNWsmaY1C8ji+SCWWP/M/uIDZHM9fi9wucHJsBY=";
  };

  nativeBuildInputs = [
    autoPatchelfHook
    dpkg
    makeWrapper
    wrapGAppsHook
  ];

  buildInputs = [
    libappindicator
    libnotify
    libsecret
    mpv-unwrapped
  ];

  dontWrapGApps = true;

  installPhase = ''
    runHook preInstall
    mkdir -p $out
    mv usr/* $out
    runHook postInstall
  '';

  preFixup = ''
    patchelf $out/share/spotube/lib/libmedia_kit_native_event_loop.so \
        --replace-needed libmpv.so.1 libmpv.so.2
  '';

  postFixup = ''
    makeWrapper $out/share/spotube/spotube $out/bin/spotube \
        "''${gappsWrapperArgs[@]}" \
        --prefix LD_LIBRARY_PATH : $out/share/spotube/lib:${lib.makeLibraryPath [ mpv-unwrapped ]} \
        --prefix PATH : ${lib.makeBinPath [ xdg-user-dirs ]}
  '';

  meta = {
    changelog = "https://github.com/KRTirtho/spotube/blob/v${finalAttrs.version}/CHANGELOG.md";
    description = "Open source Spotify client";
    longDescription = ''
      Spotube is an open source, cross-platform Spotify client that
      doesn't require Premium nor uses Electron
    '';
    homepage = "https://spotube.netlify.app/";
    license = lib.licenses.bsdOriginal;
    maintainers = with lib.maintainers; [ tomasajt ];
    platforms = [ "x86_64-linux" ];
    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
  };
})
```

In an earlier comment someone was having issues with the xdg dirs, looks like adding `xdg-user-dirs` to the path solves that issue.

**Comment by Steinhagen** [comment]
I gave it a try and it works as expected. Nice one! Do you want to create a PR for it?

**Comment by TomaSajt** [comment]
I'll try to open a PR later today.
Might need some help with supporting darwin, though.

**Comment by TomaSajt** [comment]
Opened a PR: https://github.com/NixOS/nixpkgs/pull/282975

**Comment by massix** [comment]
> Opened a PR: #282975

Hello @TomaSajt, I didn't see that you also opened a PR for Spotube, I did the same at #283156 :D we're basically doing the same thing, with the exception that you're using the debian package as a source while I'm using the binary distribution. Which one should we keep? I'd vote for yours since you're also taking care of Darwin (I don't have an OSX system so I can't check).

**Comment by TomaSajt** [comment]
>  Which one should we keep? I'd vote for yours since you're also taking care of Darwin (I don't have an OSX system so I can't check).

I'd also go with my PR, as the deb package has everything in the correct place by default.

I wonder why your derivation has much more patching than mine...

**Comment by massix** [comment]
> I'd also go with my PR, as the deb package has everything in the correct place by default.

Sold! I'll close my PR then. Thanks!


> I wonder why your derivation has much more patching than mine... Edit: I see you were not using autoPatchelfHook

Yeah, I had issues with `libmpv` using `autoPatchelfHook`, so I took the longest route :)

**Comment by TomaSajt** [comment]
Okay, so I was looking at the way upstream packages from source, and it seems like they are injecting obfuscated secrets into the generated library using [Envied](https://pub.dev/packages/envied#obfuscationencryption). Those secrets include some API keys, the spotify client IDs and secrets.
These secrets could be reverse-engineered from the binary with some effort, but I don't think the package owner would want that.
It's probably not a good idea to include them in nixpkgs itself.
So even if we found a working way of building from source, we'd never be able to simply distribute it using nixpkgs, unless we require people to supply their own client info with some thing like `spotube.override { spotifyClient = something; }`

**Comment by massix** [comment]
> Okay, so I was looking at the way upstream packages from source, and it seems like they are injecting obfuscated secrets into the generated library using [Envied](https://pub.dev/packages/envied#obfuscationencryption). Those secrets include some API keys, the spotify client IDs and secrets. These secrets could be reverse-engineered from the binary with some effort, but I don't think the package owner would want that. It's probably not a good idea to include them in nixpkgs itself. So even if we found a working way of building from source, we'd never be able to simply distribute it using nixpkgs, unless we require people to supply their own client info with some thing like `spotube.override { spotifyClient = something; }`

I reached the same conclusion while trying to build a derivation from the source code, I think that in this case using the pre built binaries is the only viable solution

**Comment by tomodachi94** [comment]
Dupe of #284081 and solved by #282975
